<a href="https://colab.research.google.com/github/oladimejiala/Schistosoma-Ova-Detect-and-Count-SODaC-Model/blob/first/Training_Parasite_Image_datasets_Using_Convluted_Neural_Networks_(CNN).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [23]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator
from tensorflow.keras.preprocessing import image
# from keras.optimizers import RMSprop
import tensorflow as tf
from tensorflow.keras.utils import to_categorical
import matplotlib.pyplot as plt
from imutils import paths
import numpy as np
import imutils
import cv2
import os
import openai
from google.colab import userdata

In [5]:
!pip install python-dotenv

In [29]:
from dotenv import load_dotenv
load_dotenv()
userdata.get('GOOGLE_API_KEY')

'AIzaSyBpPI6IMlltRi7EIh6rWGZ_ajZIskPOU2s'

In [30]:
# testing ai response from openai model
# res = userdata.chat.completions.create( # Changed openai.ChatCompletion.create to openai.chat.completions.create
#     model="gpt-3.5-turbo",
#     messages=[{"role": "user", "content": "write an alorithm that learns the structure of Schiistosomes ova and gives valid description of observed ove "}]
# )

In [14]:
from tensorflow.keras.utils import to_categorical # this is to_categorical method of one-hot encoding of label in classification models

# List of string labels
labels = ['Schistosoma haematobium', 'Schistosoma intercalatum', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi']

# Dictionary mapping string labels to integer labels
label_dict = {'Schistosoma haematobium': 0, 'Schistosoma intercalatum': 1, 'Schistosoma japonicum': 2, 'Schistosoma mansoni': 3, 'Schistosoma mekongi': 4}

# Convert string labels to integer labels
int_labels = [label_dict[label] for label in labels]

# One-hot encode integer labels
one_hot_labels = to_categorical(int_labels, num_classes=5)

print(one_hot_labels)

[[1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0.]
 [0. 0. 1. 0. 0.]
 [0. 0. 0. 1. 0.]
 [0. 0. 0. 0. 1.]]


The preprocessing step is an important part of the image classification pipeline as it can help to enhance features that distinguish the different species of Schistosoma. Here's an example of how to preprocess the images using OpenCV:

In [17]:
import cv2
import numpy as np

def preprocess_image(image_path, image_size):
    # Load image and resize to desired size
    image = cv2.imread(image_path)
    image = cv2.resize(image, image_size)

    # Convert to grayscale
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)

    # Apply histogram equalization to enhance contrast
    clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8,8))
    equalized = clahe.apply(gray)

    # Apply thresholding to binarize the image
    _, thresholded = cv2.threshold(equalized, 0, 255, cv2.THRESH_BINARY+cv2.THRESH_OTSU)

    # Apply median filtering to remove noise
    denoised = cv2.medianBlur(thresholded, 3)

    # Apply dilation and erosion to enhance features
    kernel = np.ones((3,3), np.uint8)
    dilated = cv2.dilate(denoised, kernel, iterations=1)
    eroded = cv2.erode(dilated, kernel, iterations=1)

    # Convert back to color and normalize pixel values to range [0, 1]
    processed_image = cv2.cvtColor(eroded, cv2.COLOR_GRAY2RGB)
    processed_image = processed_image.astype("float32") / 255.0

    return processed_image

# calling this function:
# preprocess_image()

The cell below, ImageDataGenerator from Keras was used to perform data augmentation, normalization, and resizing. We also define the input shape of the model, the batch size, and the paths to the train, validation, and test data directories. Finally, we use the data generators to load the images from the directories. Note that flow_from_directory method is used to load the images and their labels from the directories in batches.

In [31]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Define the data generator for training set
train_datagen = ImageDataGenerator(
    rescale=1./255,  # normalize the pixel values to range [0,1]
    rotation_range=20,  # randomly rotate the images by up to 20 degrees
    width_shift_range=0.1,  # randomly shift the images horizontally by up to 10% of the width
    height_shift_range=0.1,  # randomly shift the images vertically by up to 10% of the height
    shear_range=0.1,  # randomly shear the images by up to 10%
    zoom_range=0.1,  # randomly zoom the images by up to 10%
    horizontal_flip=True,  # randomly flip the images horizontally
    fill_mode='nearest'  # fill in any missing pixels with the nearest pixel value
)

# Define the data generator for validation and test sets
val_datagen = ImageDataGenerator(
    rescale=1./255  # normalize the pixel values to range [0,1]
)

# Define the input shape for the model
input_shape = (200, 200, 3)

# Define the batch size
batch_size = 32

# Define the paths to the train, validation, and test data directories
train_dir = r'C:\Users\Mike_Dev\Desktop\Schisto images\train'
# val_dir = r'C:\Users\Mike_Dev\Desktop\Schisto images\validation'
# test_dir = r'C:\Users\Mike_Dev\Desktop\Schisto images\test'

# Use the data generators to load the images from the directories
train_generator = train_datagen.flow_from_directory(
    train_dir,
    target_size=input_shape[:2],
    batch_size=batch_size,
    class_mode='categorical'
)

val_generator = val_datagen.flow_from_directory(
    val_dir,
    target_size=input_shape[:2],
    batch_size=batch_size,
    class_mode='categorical',
    shuffle=False
)

test_generator = val_datagen.flow_from_directory(
    test_dir,
    target_size=input_shape[:2],
    batch_size=batch_size,
    class_mode='categorical',
    shuffle=False
)

FileNotFoundError: [Errno 2] No such file or directory: 'C:\\Users\\Mike_Dev\\Desktop\\Schisto images\\train'

The code above defines data generators for the training and validation data, sets the directory paths and image size, and uses data augmentation and normalization to preprocess the images. The fit method is then used to train the model for 10 epochs using the training data and validate it using the validation data. The training history is stored in the history object, which can be used for further analysis and visualization.

In [ ]:
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv2D, MaxPooling2D, Flatten, Dense

# Define the model architecture
model = Sequential()
model.add(Conv2D(32, (3, 3), activation='relu', input_shape=(200, 200, 3)))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(64, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))
model.add(Conv2D(128, (3, 3), activation='relu'))
model.add(MaxPooling2D((2, 2)))
model.add(Flatten())
model.add(Dense(128, activation='relu'))
model.add(Dense(5, activation='softmax'))

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print the model summary
model.summary()

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d (Conv2D)              (None, 198, 198, 32)      896       
_________________________________________________________________
max_pooling2d (MaxPooling2D) (None, 99, 99, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 97, 97, 64)        18496     
_________________________________________________________________
max_pooling2d_1 (MaxPooling2 (None, 48, 48, 64)        0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 46, 46, 128)       73856     
_________________________________________________________________
max_pooling2d_2 (MaxPooling2 (None, 23, 23, 128)       0         
____

In [ ]:
from tensorflow.keras.preprocessing.image import ImageDataGenerator

# Define data generators for training and validation data
train_datagen = ImageDataGenerator(
        rescale=1./255,
        shear_range=0.2,
        zoom_range=0.2,
        horizontal_flip=True)

val_datagen = ImageDataGenerator(rescale=1./255)

# Set the directory paths for the training and validation data
# train_dir = 'path/to/train/data'
# val_dir = 'path/to/validation/data'

# Set the image size and batch size for training and validation data
img_size = (200, 200)
batch_size = 16

# Generate the training and validation data batches
train_generator = train_datagen.flow_from_directory(
        train_dir,
        target_size=img_size,
        batch_size=batch_size,
        class_mode='categorical')

validation_generator = val_datagen.flow_from_directory(
        val_dir,
        target_size=img_size,
        batch_size=batch_size,
        class_mode='categorical')

# Train the model
history = model.fit(
        train_generator,
        steps_per_epoch=train_generator.samples // batch_size,
        epochs=40,
        validation_data=validation_generator,
        validation_steps=validation_generator.samples // batch_size)


Found 64 images belonging to 5 classes.
Found 64 images belonging to 5 classes.
Epoch 1/40
4/4 [==============================] - 19s 5s/step - loss: 2.1304 - acc: 0.2188 - val_loss: 1.4571 - val_acc: 0.4375
Epoch 2/40
4/4 [==============================] - 8s 2s/step - loss: 1.4466 - acc: 0.3594 - val_loss: 1.2511 - val_acc: 0.4375
Epoch 3/40
4/4 [==============================] - 7s 2s/step - loss: 1.2626 - acc: 0.5156 - val_loss: 1.2732 - val_acc: 0.3906
Epoch 4/40
4/4 [==============================] - 8s 2s/step - loss: 1.3484 - acc: 0.3281 - val_loss: 1.1783 - val_acc: 0.5000
Epoch 5/40
4/4 [==============================] - 8s 2s/step - loss: 1.2389 - acc: 0.3750 - val_loss: 1.1794 - val_acc: 0.4062
Epoch 6/40
4/4 [==============================] - 8s 2s/step - loss: 1.2302 - acc: 0.4062 - val_loss: 1.0864 - val_acc: 0.5000
Epoch 7/40
4/4 [==============================] - 8s 2s/step - loss: 1.1089 - acc: 0.5781 - val_loss: 1.0161 - val_acc: 0.5781
Epoch 8/40
4/4 [==============

In [ ]:
# Train the model on the preprocessed data
history = model.fit(train_generator, epochs=10, validation_data=val_generator)

# Evaluate the model on the test data
test_loss, test_acc = model.evaluate(test_generator)

model.save(r"C:\\Users\Mike_Dev\Desktop\Schisto images\schistosoma_ova_classifier.h5")

# Print the test accuracy
print(f"Test accuracy: {test_acc:.4f}")
print(f"Test Loss: {test_loss:.4f}")

Epoch 1/10
4/4 [==============================] - 29s 7s/step - loss: 0.3221 - acc: 0.9531 - val_loss: 0.0800 - val_acc: 0.9688
Epoch 2/10
4/4 [==============================] - 29s 7s/step - loss: 0.1486 - acc: 0.9375 - val_loss: 0.0939 - val_acc: 0.9688
Epoch 3/10
4/4 [==============================] - 28s 7s/step - loss: 0.0583 - acc: 0.9844 - val_loss: 0.0687 - val_acc: 0.9688
Epoch 4/10
4/4 [==============================] - 28s 7s/step - loss: 0.0682 - acc: 0.9844 - val_loss: 0.0518 - val_acc: 0.9844
Epoch 5/10
4/4 [==============================] - 28s 7s/step - loss: 0.0652 - acc: 0.9844 - val_loss: 0.0698 - val_acc: 0.9688
Epoch 6/10
4/4 [==============================] - 30s 8s/step - loss: 0.0612 - acc: 0.9688 - val_loss: 0.0793 - val_acc: 0.9688
Epoch 7/10
4/4 [==============================] - 27s 7s/step - loss: 0.0732 - acc: 0.9531 - val_loss: 0.0346 - val_acc: 0.9844
Epoch 8/10
4/4 [==============================] - 28s 7s/step - loss: 0.0330 - acc: 1.0000 - val_loss: 0

From above:
first run: A test accuracy of 0.953125 and test loss of 0.1164017990231514 indicate that the model is performing very well on the test set. It means that the model correctly classified 95.31% of the test data and has a low error rate of 0.116. These are good performance metrics, indicating that the model is able to generalize well to unseen data.

Third run: using f-strings gives
Test accuracy: 0.9844
Test Loss: 0.0454

In the model evaluation cell below, only one image data was used to evaluate the model which gives 0.2000 i.e 20% accuracy of the while the one above gave 98.44% based on numerous test data used.

In [ ]:
# # Evaluate the model on the test set
# from tensorflow.keras.utils import to_categorical
# from sklearn.preprocessing import LabelEncoder, OneHotEncoder

# label_encoder = LabelEncoder()
# # processed_image = preprocess_image(image_path=r"C:\Users\Mike_Dev\Desktop\Schisto images\train", image_size=(200, 200))
# test_data = preprocess_image(r"C:\Users\Mike_Dev\Desktop\Schistosoma mansoni.jpg", (200, 200))[np.newaxis, ...]
# test_data = np.repeat(test_data, 5, axis=0)
# test_labels = ['Schistosoma haematobium', 'Schistosoma intercalatum', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi'] # assign the appropriate label to the test_data

# # encode the labels as one-hot vectors
# test_labels_encoded = to_categorical(label_encoder.fit_transform(test_labels))

# # evaluate the model on the test set
# loss, accuracy = model.evaluate(test_data, test_labels_encoded)
# print(f"Test Loss: {loss:.4f}")
# print(f"Test Accuracy: {accuracy:.4f}")


# # loss, accuracy = model.evaluate(np.array([test_data]), np.array([test_labels]))
# # print(f"Test Loss: {loss:.4f}")
# # print(f"Test Accuracy: {accuracy:.4f}")

# """
# You can also use the predict() method of the model to get the predicted
# labels for the test set, and compare them with the true labels to get a
# more detailed performance analysis.
# """

In [ ]:
# # alternatively using the pretrained VGG model.
# from tensorflow.keras.applications.vgg16 import VGG16
# model = VGG16()

In [ ]:
# from tensorflow.keras.preprocessing import image
# from tensorflow.keras.applications.vgg16 import preprocess_input

# img = image.load_img('path/to/image.jpg', target_size=(224, 224))
# x = image.img_to_array(img)
# x = preprocess_input(x)

# import numpy as np

# preds = model.predict(np.array([x]))

Integrating the Convolutional Neural Network Model into Schistosoma ova detection Function. - This is the first one today, but seem not working.

In [ ]:
# import cv2
# import tensorflow as tf

# class SchistosomaOvaCounter(tf.keras.layers.Layer):
#     def __init__(self):
#         super(SchistosomaOvaCounter, self).__init__()

#     def call(self, inputs):
#         image_path = inputs.numpy()[0].decode()

#         # Load the image
#         image = cv2.imread(image_path)

#         # Predict the species of Schistosoma based on the filename
#         if "haematobium" in image_path.lower():
#             species = "haematobium"
#             ovum_mask = predict_haematobium(image)
#         elif "mansoni" in image_path.lower():
#             species = "mansoni"
#             ovum_mask = predict_mansoni(image)
#         elif "intercalatum" in image_path.lower():
#             species = "intercalatum"
#             ovum_mask = predict_intercalatum(image)
#         elif "japonicum" in image_path.lower():
#             species = "japonicum"
#             ovum_mask = predict_japonicum(image)
#         elif "mekongi" in image_path.lower():
#             species = "mekongi"
#             ovum_mask = predict_mekongi(image)
#         else:
#             lab_image = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
#             l, a, b = cv2.split(lab_image)
#             clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=(8,8))
#             l = clahe.apply(l)
#             image = cv2.merge([l,a,b])
#             hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
#             green_mask = cv2.inRange(hsv_image, (36, 25, 25), (86, 255,255))
#             ovum_mask = cv2.bitwise_not(green_mask)

#         # Use the pretrained model to predict the presence of ova
#         resized_image = cv2.resize(image, (200, 200))
#         resized_image = cv2.cvtColor(resized_image, cv2.COLOR_BGR2RGB)
#         resized_image = resized_image.astype('float32') / 255.0
#         prediction = model.predict(tf.expand_dims(resized_image, axis=0), steps=1)[0]
#         has_ova = prediction[1] > prediction[0]

#         # Count the number of ovum of the predicted species in the image
#         if has_ova:
#             lab_image = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
#             ovum_mask = cv2.medianBlur(ovum_mask, 5)
#             ovum_mask = cv2.morphologyEx(ovum_mask, cv2.MORPH_OPEN, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (9, 9)))
#             ovum_mask = cv2.morphologyEx(ovum_mask, cv2.MORPH_CLOSE, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (15, 15)))
#             contours, _ = cv2.findContours(ovum_mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
#             num_ovum = len(contours)
#         else:
#             num_ovum = 0

#         return tf.constant([num_ovum], dtype=tf.float32), tf.constant([species], dtype=tf.string)

# # Load the pretrained model
# # model = tf.keras.models.load_model('schistosoma_ova_classifier.h5')

In [ ]:
# Define a Keras model with SchistosomaOvaCounter as its output layer
# input_tensor = tf.keras.Input(shape=(1,), dtype=tf.string)
# output_tensor = SchistosomaOvaCounter()(input_tensor)
# model = tf.keras.Model(inputs=input_tensor, outputs=output_tensor)

# # Count the number of ovum in the image and print the result
# image_path = 'test_image.jpg'
# num_ovum, species = model.predict(np.array([image_path]))
# print(f"There are {num_ovum} Schistosoma {species} ova in the image.")

# # Create the input prompt for the user
# print("Enter the path to the image file:")
# image_path = input()

# # Count the number of ovum in the image and print the result
# num_ovum, species = SchistosomaOvaCounter(image_path)
# print(f"There are {num_ovum} Schistosoma {species} ova in the image.")

In [ ]:
# image_path = "/path/to/your/image.jpg"
# image_tensor = tf.constant([image_path], dtype=tf.string)
# schisto_ova_counter = SchistosomaOvaCounter()
# num_ovum, species = schisto_ova_counter(image_tensor)

Integrating the Convolutional Neural Network Model into Schistosoma ova detection Function. - This is the second one today: 1/04/2023.

In [ ]:
# # Import necessary libraries
# import tensorflow as tf
# import numpy as np
# import cv2

# # Define SchistosomaOvaCounter class
# class SchistosomaOvaCounter(tf.keras.Model):

#     def __init__(self):
#         super(SchistosomaOvaCounter, self).__init__()
#         self.model = tf.keras.models.load_model('schistosoma_model.h5')
#         self.species_mapping = {0: 'S. haematobium', 1: 'S. mansoni'}

#     def call(self, inputs):
#         image_path = inputs.numpy()[0].decode()

#         # Load the image
#         img = cv2.imread(image_path)
#         img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
#         img = cv2.resize(img, (224, 224)) / 255.0
#         img = np.expand_dims(img, axis=0)

#         # Use the loaded model to make predictions
#         predictions = self.model.predict(img)
#         species_id = np.argmax(predictions)
#         num_ovum = int(np.round(predictions[0][species_id] * 100, 0))
#         species = self.species_mapping[species_id]

#         return num_ovum, species

# # Pass image path as tensor to call method
# image_path = tf.constant([r"C:\Users\Mike_Dev\Desktop\Schistosoma mansoni.jpg"])
# image_path = tf.squeeze(image_path).numpy()
# image_path = tf.strings.unicode_decode(image_path, 'UTF-8')[0]

# # Create an instance of the SchistosomaOvaCounter class
# counter = SchistosomaOvaCounter()

# # Call the `call` method of the `SchistosomaOvaCounter` class on the image tensor
# num_ovum, species = counter.call(tf.io.read_file(image_path))

# # Print the results
# print(f"Number of Schistosoma ova in the image: {num_ovum}%")
# print(f"The species of Schistosoma in the image is: {species}")


In [ ]:
# import tensorflow as tf

# # Define the SchistosomaOvaCounter class
# class SchistosomaOvaCounter(tf.keras.Model):
#     def __init__(self):
#         super(SchistosomaOvaCounter, self).__init__()
#         # Define the layers of the model

#     def call(self, inputs):
#         # Load the image
#         image = tf.io.read_file(inputs)
#         # Decode the image
#         image = tf.io.decode_jpeg(image)
#         # Resize the image
#         image = tf.image.resize(image, [224, 224])
#         # Preprocess the image
#         image = tf.keras.applications.mobilenet_v2.preprocess_input(image)
#         # Pass the image through the model

#         # Return the number of ova and species


# # Pass image path as tensor to call method
# image_path = tf.constant([r"C:\Users\Mike_Dev\Desktop\Schistosoma mansoni.jpg"])
# with tf.compat.v1.Session() as sess:
#     image_path = tf.squeeze(image_path).eval(session=sess).decode()
# # Create an instance of the SchistosomaOvaCounter class
# counter = SchistosomaOvaCounter()
# # Call the `call` method of the `SchistosomaOvaCounter` class on the image tensor
# num_ovum, species = counter.call(image_path)
# # Print the results
# print(f"Number of Schistosoma mansoni ova detected: {num_ovum}")
# print(f"Schistosoma mansoni species identified: {species}")

In [ ]:
# import tensorflow as tf
# from tensorflow.keras.applications.resnet50 import ResNet50, preprocess_input
# from tensorflow.keras.layers import Dense
# from tensorflow.keras.models import Model
# from tensorflow.keras.preprocessing import image
# import numpy as np

# class SchistosomaOvaCounter(Model):
#     def __init__(self):
#         super(SchistosomaOvaCounter, self).__init__()
#         self.resnet = ResNet50(weights='imagenet', include_top=False, input_shape=(224, 224, 3))
#         self.pool = tf.keras.layers.GlobalAveragePooling2D()
#         self.fc1 = tf.keras.layers.Dense(1024, activation='relu')
#         self.fc2 = tf.keras.layers.Dense(1, activation='linear')

#     def call(self, inputs):
#         x = preprocess_input(inputs)
#         x = tf.image.resize(x,[200,200])
#         x = self.resnet(x)
#         x = self.pool(x)
#         x = self.fc1(x)
#         num_ovum = self.fc2(x)
#         return num_ovum


# # Load image using tf.io.read_file and decode using tf.image.decode_jpeg
# image_path = tf.constant([r"C:\Users\Mike_Dev\Desktop\Schistosoma mansoni.jpg"])
# image = tf.io.read_file(tf.squeeze(image_path))
# image = tf.image.decode_jpeg(image, channels=3)

# # Create an instance of the SchistosomaOvaCounter class
# counter = SchistosomaOvaCounter()

# # Call the `call` method of the `SchistosomaOvaCounter` class on the image tensor
# num_ovum = counter(tf.expand_dims(image, axis=0))

# # Print the results
# with tf.compat.v1.Session() as sess:
#     num_ovum_value = sess.run(num_ovum)

# print(f"Number of Schistosoma mansoni ova detected: {num_ovum_value[0][0]:.2f}")

In [ ]:
# from tensorflow.keras.applications.resnet50 import ResNet50
# from tensorflow.keras.layers import Dense, Flatten
# from tensorflow.keras.models import Model

# # Load the pre-trained ResNet50 model
# resnet = ResNet50(weights='imagenet', include_top=False, input_shape=(200, 200, 3))

# # Define the new architecture
# x = Flatten()(resnet.output)
# x = Dense(256, activation='relu')(x)
# x = Dense(128, activation='relu')(x)
# output = Dense(10, activation='softmax')(x)
# model = Model(inputs=resnet.input, outputs=output)

# # Compile the new model
# model.compile(loss='categorical_crossentropy', optimizer='adam', metrics=['accuracy'])

# # Train the
# model.fit(
#         train_generator,
#         steps_per_epoch=train_generator.samples // batch_size,
#         epochs=40,
#         validation_data=validation_generator,
#         validation_steps=validation_generator.samples // batch_size)

# # Evaluate the model
# loss, accuracy = model.evaluate(test_dir, labels)

# # Save the new model
# model.save('new_resnet_model.h5')


In [ ]:
# import cv2
# import numpy as np
# import tensorflow as tf

# def load_model():
#     model_path = "schisto_ova_classifier.h5"
#     model = tf.keras.models.load_model(model_path)
#     return model

# def predict_species(image):
#     model = load_model()
#     image = cv2.resize(image, (128, 128))
#     image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
#     image = np.expand_dims(image, axis=0)
#     prediction = model.predict(image)
#     predicted_class = np.argmax(prediction)
#     if predicted_class == 0:
#         return "haematobium"
#     elif predicted_class == 1:
#         return "mansoni"
#     else:
#         return None

# def predict_haematobium(image):
#     # code to predict ovum mask for haematobium species
#     pass

# def predict_mansoni(image):
#     # code to predict ovum mask for mansoni species
#     pass

# def schistosoma_ova_counter(image_path):
#     # Load the image
#     image = cv2.imread(image_path)

#     # Predict the species of Schistosoma based on the filename
#     if "haematobium" in image_path.lower():
#         species = "haematobium"
#         ovum_mask = predict_haematobium(image)
#     elif "mansoni" in image_path.lower():
#         species = "mansoni"
#         ovum_mask = predict_mansoni(image)
#     else:
#         lab_image = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
#         l, a, b = cv2.split(lab_image)
#         clahe = cv2.createCLAHE(clipLimit=3.0, tileGridSize=(8, 8))
#         l = clahe.apply(l)
#         lab_image = cv2.merge((l, a, b))
#         image = cv2.cvtColor(lab_image, cv2.COLOR_LAB2BGR)
#         species = predict_species(image)
#         if species == "haematobium":
#             ovum_mask = predict_haematobium(image)
#         elif species == "mansoni":
#             ovum_mask = predict_mansoni(image)
#         else:
#             return "Unable to identify Schistosoma species from image"

#     # Count the number of ovum of the predicted species in the image
#     lab_image = cv2.cvtColor(image, cv2.COLOR_BGR2LAB)
#     ovum_mask = cv2.medianBlur(ovum_mask, 5)
#     ovum_mask = cv2.morphologyEx(ovum_mask, cv2.MORPH_OPEN, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (9, 9)))
#     contours, hierarchy = cv2.findContours(ovum_mask, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
#     num_ovum = len(contours)

#     return f"{num_ovum} Schistosoma {species} ovum found in the image"


This output of the cell below indicates that the CNN model identified Schistosoma mansoni with a probability of 0.997 and Schistosoma japonicum with a probability of 0.0028 in the input image. The probabilities of the other species (Schistosoma haematobium, Schistosoma intercalatum, and Schistosoma mekongi) were 0.0. With 99.7% of S.mansoni and 0.28& of S.japonicum, this could tell of potential HYBRID EGG.

In [ ]:
import tensorflow as tf
import numpy as np
import cv2

# Load the pre-trained CNN model
model = tf.keras.models.load_model('schistosoma_ova_classifier.h5')

# Preprocess the image
img = cv2.imread(r"C:\Users\Mike_Dev\Desktop\sh and sm.jpg")
img = cv2.resize(img, (200, 200))
img = np.expand_dims(img, axis=0)

# Pass the image through the CNN model
preds = model.predict(img)

# Postprocess the output
pred_labels = np.argmax(preds, axis=1)
counts = np.sum(preds, axis=0)

# Return the results
species = ['Schistosoma haematobium', 'Schistosoma intercalatum', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi']
results = dict(zip(species, counts))
print(results)

{'Schistosoma haematobium': 0.0, 'Schistosoma intercalatum': 0.0, 'Schistosoma japonicum': 0.0028280779, 'Schistosoma mansoni': 0.99717194, 'Schistosoma mekongi': 0.0}


Schisto detect and count function

In [ ]:
import tensorflow as tf
import numpy as np
import cv2

# Define the function
def detect_and_count_ova(image_path, model_path):
    # Load the pre-trained CNN model
    model = tf.keras.models.load_model(model_path)

    # Preprocess the image
    img = cv2.imread(image_path)
    img = cv2.resize(img, (200, 200))
    img = np.expand_dims(img, axis=0)

    # Pass the image through the CNN model
    preds = model.predict(img)

    # Postprocess the output
    pred_labels = np.argmax(preds, axis=1)
    counts = np.sum(preds, axis=0)

    # Define the species labels
    species = ['Schistosoma haematobium', 'Schistosoma intercalatum', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi']

    # Get the species counts and labels
    species_counts = dict(zip(species, counts))
    identified_species = [species[i] for i in np.where(pred_labels > 0)[0]]

    # Return the results
    return identified_species, species_counts

this one identified S. haem and counted S.mansoni :
Identified species: ['Schistosoma haematobium']
Species counts: {'Schistosoma haematobium': 0.0, 'Schistosoma intercalatum': 0.0, 'Schistosoma japonicum': 0.0, 'Schistosoma mansoni': 1.0, 'Schistosoma mekongi': 0.0}

In [ ]:
identified_species, species_counts = detect_and_count_ova(r"C:\Users\Mike_Dev\Desktop\multiple sm.jpg", 'schistosoma_ova_classifier.h5')
print('Identified species:', identified_species)
print('Species counts:', species_counts)

Identified species: ['Schistosoma haematobium']
Species counts: {'Schistosoma haematobium': 0.0, 'Schistosoma intercalatum': 0.0, 'Schistosoma japonicum': 0.0, 'Schistosoma mansoni': 1.0, 'Schistosoma mekongi': 0.0}


In [ ]:
import tensorflow as tf
import numpy as np
import cv2

# Define the function
def detect_and_count_ovum(image_path, model_path, threshold=0.5):
    # Load the pre-trained CNN model
    model = tf.keras.models.load_model(model_path)

    # Preprocess the image
    img = cv2.imread(image_path)
    img = cv2.resize(img, (200, 200))
    img = np.expand_dims(img, axis=0)

    # Pass the image through the CNN model
    preds = model.predict(img)

    # Postprocess the output
    pred_labels = np.where(preds > threshold)[1]
    counts1 = np.sum(preds[:, pred_labels], axis=0)

    # Define the species labels
    species1 = ['Schistosoma haematobium', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi', 'Schistosoma intercalatum']

    # Get the species counts and labels
    species_counts1 = dict(zip(species, counts1))
    identified_species1 = [species[i] for i in pred_labels]

    # Return the results
    return identified_species1, species_counts1


In [ ]:
identified_species, species_counts = detect_and_count_ovum(r"C:\Users\Mike_Dev\Desktop\multiple sm.jpg", 'schistosoma_ova_classifier.h5')
print('Identified species:', identified_species)
print('Species counts:', species_counts)

Identified species: ['Schistosoma mansoni']
Species counts: {'Schistosoma haematobium': 1.0}


In [ ]:
import tensorflow as tf
import numpy as np
import cv2

def detect_and_count_ova(image_path, model_path):
    # Load the pre-trained CNN model
    model = tf.keras.models.load_model(model_path)

    # Preprocess the image
    img = cv2.imread(image_path)
    img = cv2.resize(img, (200, 200))
    img = np.expand_dims(img, axis=0)

    # Pass the image through the CNN model
    preds = model.predict(img)

    # Define the species labels
    species = ['Schistosoma haematobium', 'Schistosoma japonicum', 'Schistosoma mansoni', 'Schistosoma mekongi', 'Schistosoma intercalatum']

    # Define the morphology of each ovum
    morphology = {'Schistosoma haematobium': 'large terminal spine',
                  'Schistosoma japonicum': 'small lateral spine',
                  'Schistosoma mansoni': 'large lateral spine',
                  'Schistosoma mekongi': 'large terminal spine',
                  'Schistosoma intercalatum': 'small terminal spine'}

    # Count the number of ova detected per species
    counts = {}
    for i in range(len(preds[0])):
        if preds[0][i] > 0.5:
            species_name = species[i]
            if species_name in counts:
                counts[species_name] += 1
            else:
                counts[species_name] = 1

    # Calculate the total number of ova detected
    total_count = sum(counts.values())

    # Return the results
    return counts, morphology, total_count

In [ ]:
counts, morphology, total_count = detect_and_count_ova(r"C:\Users\Mike_Dev\Desktop\multiple sm.jpg", 'schistosoma_ova_classifier.h5')
print('Counts:', counts)
print('Morphology:', morphology)
print('Total count:', total_count)

Counts: {'Schistosoma mekongi': 1}
Morphology: {'Schistosoma haematobium': 'large terminal spine', 'Schistosoma japonicum': 'small lateral spine', 'Schistosoma mansoni': 'large lateral spine', 'Schistosoma mekongi': 'large terminal spine', 'Schistosoma intercalatum': 'small terminal spine'}
Total count: 1


Trying the Masked R-CNN in place of CNN that i've been using all the while without satisfactory results.

In [ ]:
conda install tensorflow